ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ACSReader.scala

```diff
@@ -107,32 +119,97 @@ class FilterTableACSReader(
       .map(toIdSource)
       .pipe(mergeSort[Long])
       .statefulMapConcat(statefulDeduplicate)
-      .grouped(pageSize)
-      .map(_.toVector)
+      .via(
+        BatchN(
+          maxBatchSize = pageSize,
+          maxBatchCount = acsFetchingparallelism + 1,
+        )
+      )
       .async
+      .addAttributes(
+        Attributes.inputBuffer(initial = acsFetchingparallelism, max = acsFetchingparallelism)
+      )
       .mapAsync(acsFetchingparallelism)(fetchAcs)
   }
 }
 
 private[events] object FilterTableACSReader {
+  private val logger = ContextualizedLogger.get(this.getClass)
+
   case class Filter(party: Party, templateId: Option[Identifier])
 
+  case class IdQuery(fromExclusiveEventSeqId: Long, pageSize: Int)
+
+  case class IdQueryConfiguration(
+      minPageSize: Int,
+      maxPageSize: Int,
+  )
+
+  object IdQueryConfiguration {
+    def apply(
+        maxIdPageSize: Int,
+        idPageWorkingMemoryBytes: Int,
+        filterSize: Int,
+        idPageBufferSize: Int,
+    )(implicit loggingContext: LoggingContext): IdQueryConfiguration = {
+      val LowestIdPageSize = 10
+      // Approximation how many index entries can be present in one btree index leaf page (fetching smaller than this only adds round-trip overhead, without boiling down to smaller disk read)
+      // Experiments show party_id, template_id index has 244 tuples per page, wildcard party_id index has 254 per page (with default fill ratio for BTREE Index)
+      // Picking a lower number is for accommodating pruning, deletions, index bloat effect, which boil down to lower tuple per page ratio.
+      val RecommendedMinIdPageSize = 200
+      val calculatedMaxIdPageSize =
+        idPageWorkingMemoryBytes
+          ./(8) // IDs stored in 8 bytes
+          ./(
+            idPageBufferSize + 1
```

Comment:
Why is there +1 ?

Reply:
If the stream is stabilized (initialized, and the backpressure is exerted from either 2nd phase or the consumer) one will be kept in memory in order to feed the merge sort network. and idPageBufferSize will be in the buffer